Paul Kilgarriff - Weekend 3 Challenge

PKilgarriff opened the pull request:
# Paul Kilgarriff

Concerned about the Test Coverage, as SimpleCov started just giving me 100% even when I actively commented out entire tests for multiple methods. Otherwise, the actual game logic was straightforward, but as expected I found it tricky to know when and how to encapsulate methods away from parts of the MVC.

# User stories 

Please list which user stories you've implemented (delete the ones that don't apply).

- [x] User story 1: "I would like to register my name before playing an online game"
- [x] User story 2: "I would like to be able to play rock/paper/scissors"

# README checklist

Does your README contains instructions for

- [x] how to install,
- [x] how to run,
- [x] and how to test your code?

[Here is a pill](https://github.com/makersacademy/course/blob/main/pills/readmes.md) that can help you write a great README!

PKilgarriff committed:
Add Game and Player classes

Test driven development for that basic Player and Game class
Views created for index, name_entry, and welcome

Rackup, Rubocop, and Sinatar configured

PKilgarriff committed:
Update Game class

Handle computer controlled players (defaults to User controlled)

PKilgarriff committed:
Update Player class

- add set_choice method
- add clear_choice method
- add attr_reader for choice

Unit test for the above methods

PKilgarriff committed:
Add feature test for User Story 1

Update spec_helper to configure Capybara for automated browser testing

Check that app can store and display a user name

PKilgarriff committed:
Update Game class

- Add add_computer_opponents method to remove function from controller
- add compare_choices method that currently outputs  player choices formatted

PKilgarriff committed:
Add process model for RPSApp

PKilgarriff committed:
Update Player class

- add choice_made? method
- add set_random_choice method

PKilgarriff committed:
Update Game class

- update compare_choices method to return a string of the outcome
- add winner instance variable

PKilgarriff committed:
Update controller and views

- add routes for
  - rules
  - game-screen
  - player selection pages
  - name-entry

PKilgarriff committed:
Update Game class

- Rename compare_choices to outcome
- refactor method for clarity

PKilgarriff committed:
Update views

Add body tags to all files

PKilgarriff committed:
Add GameHelper module

Include setup_1_player and setup_2_players methods for use in spec files

PKilgarriff committed:
Update Game class

- add choices_made? method to remove logic from controller
- add reset_choices method
  - clears player choices without wiping players array
- extract logic from outcome method into player1_win? private method

PKilgarriff committed:
Update controller

- add post route for name-entry instead of number-of-players
- enable sessions to pass number of players between post and get routes
- add choice updates to the player-choice routes

PKilgarriff committed:
Tidy up after running Rubocop

- Add labels to form on name_entry view
- change variable names to normalcase with numbers

PKilgarriff committed:
Add Process Diagram as HTML file

PKilgarriff committed:
Add head to views

- Add titles to each view
- also redirected play-again towards game instead of welcome screen

PKilgarriff committed:
Update controller

- handle case where Players don't enter any names
- remove unnecessary Number of players line from welcome

PKilgarriff committed:
Update travis config file

Require ruby version 3.0.2

PKilgarriff committed:
Correct spelling of Marketeer

PKilgarriff committed:
Add pull request template

PKilgarriff committed:
Switch PK-notes to Readme for Code Review

Lorn888 commented on README.md:
```diff
@@ -1,21 +1,80 @@
 # RPS Challenge
 
```
Well written and detailed README file.

Lorn888 commented on spec/units/game_spec.rb:
```diff
@@ -0,0 +1,57 @@
+require 'game'
```
Feature tests seem to cover every scenario.

Lorn888 commented on views/game_complete.erb:
```diff
@@ -0,0 +1,13 @@
+<html>
+  <head>
+    <title>RPS - Finished</title>
```
Added titles to each view is a nice touch.

Lorn888 reviewed (commented):
Well written code that makes accessing the game and its features intuitive and user friendly. A+

PKilgarriff committed:
Update Gemfile to include sinatra-contrib

PKilgarriff committed:
Run bundle install after changing Gemfile

PKilgarriff committed:
Move SimpleCov config to top of spec_helper

Solves problem with incorrect test coverage

PKilgarriff committed:
Update spec files to increase test coverage

Previously misconfigured SimpleCov - inaccurately reported 100% coverage

- SimpleCov config moved in spec_helper
- missing tests written

PKilgarriff committed:
Update Game class

-add create_players method to move business logic from controller
